src/sinol_make/commands/gen/__init__.py

```diff
@@ -0,0 +1,147 @@
+                        'is a new test or it differs from the previous one, '
+                        'then the output file will be generated using the '
+                        'correct solution. You can also specify your ingen source '
+                        'file which will be used.'
+        )
+
+        parser.add_argument('ingen_path', type=str, nargs='?',
+                            help='path to ingen source file, for example prog/abcingen.cpp')
+        parser.add_argument('-c', '--cpus', type=int,
+                            help=f'number of cpus to use, by default {mp.cpu_count()} (all available). '
+                                 f'Used when generating output files.', default=mp.cpu_count())
+        parsers.add_compilation_arguments(parser)
+
+    def compile_ingen(self):
+        self.ingen_exe, compile_log_path = gen_util.compile_ingen(self.ingen, self.args,
+                                                                  self.args.weak_compilation_flags)
+        if self.ingen_exe is None:
+            print(util.error('Failed ingen compilation.'))
+            compile.print_compile_log(compile_log_path)
+            exit(1)
+        else:
+            print(util.info('Successfully compiled ingen.'))
+
+    def compile_correct_solution(self):
+        self.correct_solution_exe, compile_log_path = gen_util.compile_correct_solution(self.correct_solution, self.args,
+                                                                                        self.args.weak_compilation_flags)
+        if self.correct_solution_exe is None:
+            print(util.error('Failed ingen compilation.'))
+            compile.print_compile_log(compile_log_path)
+            exit(1)
+        else:
+            print(util.info('Successfully compiled ingen.'))
+
+    def generate_outputs(self, outputs_to_generate):
+        print(f'Generating output files for {len(outputs_to_generate)} tests on {self.args.cpus} cpus.')
+        arguments = []
+        for output in outputs_to_generate:
+            output_basename = os.path.basename(output)
+            input = os.path.join(os.getcwd(), 'in', os.path.splitext(output_basename)[0] + '.in')
+            arguments.append(OutputGenerationArguments(self.correct_solution_exe, input, output))
+
+        failed = False
+        with mp.Pool(self.args.cpus) as pool:
+            for i, result in enumerate(pool.imap_unordered(gen_util.generate_output, arguments)):
+                if result:
+                    print(util.info(f'Successfully generated output file {os.path.basename(arguments[i].output_test)}'))
+                else:
+                    failed = True
+                    print(util.error(f'Failed to generate output file {os.path.basename(arguments[i].output_test)}'))
+
+        if failed:
+            util.exit_with_error('Failed to generate some output files.')
+        else:
+            print(util.info('Successfully generated all output files.'))
```

Comment:
I'm not really a fan of the "creating a bool, changing it in a for statement, checking its results after the for" pattern, which shows up really often in the whole sinol-make's current codebase. How easy would it be to change it into a functional approach? E.g. by first creating the list of tests, then transforming it (through e.g. a `map` function, but I'm not sure whether it exists in Python) to a list of bools (that contain information whether the test was generated successfully), then checking whether a list contains only True bools (I remember a `all(list)` in Python, but there might be other ways) and then matching on the result (through a `if` or `match`)? Python might be not the best language for implementing this specific approach, but I think that there can be some other good approaches besides the current one (it's worth to spend some time googling this subject). I'm bringing this up, because it's easy to forget somewhere a variable assignment and it also increases the amount of variables in the environment, which increases the needed context to read the code, so I think this pattern is bug-prone.

Reply:
I changed it a bit, but I don't think it's really possible in this case, because `Pool.imap` executes functions in order, and returns result as soon as possible. If we completely removed printing in the for then nothing would print for a long time.

Reply:
You can still print in the for. That can be done by mapping each result, printing a message when needed, and returning a bool. After that, you check whether all bools returned true through e.g. the `all` function. What do you think of it?